src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -86,5 +113,258 @@ private protected IReadOnlyList<PackageReference> GetTransitivePackageReferences
+        {
+            token.ThrowIfCancellationRequested();
+
+            LockFileTargetLibrary node = graph
+                .Libraries
+                .Where(lib => string.Equals(lib.Name, current.PackageIdentity.Id, StringComparison.OrdinalIgnoreCase)
+                        && current.AllowedVersions.Satisfies(lib.Version)
+                        && lib.Type == LibraryType.Package.Value)
+                .FirstOrDefault();
+
+            if (node != default)
+            {
+                visited[current.PackageIdentity] = true; // visited
+
+                // Update cache
+                TransitiveEntry cachedEntry = GetCachedTransitiveOrigin(current.PackageIdentity);
+                if (cachedEntry == null)
+                {
+                    cachedEntry = new Dictionary<FrameworkRuntimePair, IList<PackageReference>>
+                    {
+                        [fxRidEntry] = new List<PackageReference>()
+                    };
+
+                }
+                if (!cachedEntry.ContainsKey(fxRidEntry))
+                {
+                    cachedEntry[fxRidEntry] = new List<PackageReference>();
+                }
+                if (!cachedEntry[fxRidEntry].Contains(top))
+                {
+                    cachedEntry[fxRidEntry].Add(top);
+                }
+                SetCachedTransitiveOrigin(current.PackageIdentity, cachedEntry);
+
+                foreach (PackageDependency dep in node.Dependencies)
+                {
+                    // Create PackageReference object as a data-model based on dependency
+                    var pkgChild = new PackageReference(
+                        identity: new PackageIdentity(dep.Id, dep.VersionRange.MinVersion),
+                        targetFramework: fxRidEntry.Framework,
+                        userInstalled: false,
+                        developmentDependency: false,
+                        requireReinstallation: false,
+                        allowedVersions: dep.VersionRange);
+
+                    if (!visited.ContainsKey(pkgChild.PackageIdentity))
+                    {
+                        MarkTransitiveOrigin(top, pkgChild, graph, visited, fxRidEntry, token);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Generates a cache key for Transitive Originas cache
+        /// </summary>
+        /// <param name="transitivePackage"></param>
+        /// <returns>A string with given key</returns>
+        /// <seealso cref="GetCachedTransitiveOrigin(PackageIdentity)"/>
+        /// <seealso cref="SetCachedTransitiveOrigin(PackageIdentity, TransitiveEntry)"/>
+        internal string GetTransitiveCacheKey(PackageIdentity transitivePackage)
+        {
+            return _projectUniqueName + "/" + transitivePackage.Id.ToLowerInvariant() + "." + transitivePackage.Version.ToNormalizedString();
```

Comment:
Same bug mentioned [here](https://github.com/NuGet/NuGet.Client/pull/4381/files) by @zivkan exist here.
I'm including screenshot since link for resolved discussion doesn't work.
![image](https://user-images.githubusercontent.com/8766776/149427035-55a51928-e4c4-4bcc-98cd-dce0e08c9185.png)

Use `-` instead of `.` or consider using same technique as @jeffkl recommended there as best perf gain (PackageIdentity comparer).

Reply:
My.Thing.21.2.3
My.Thing2.1.2.3

Reply:
Your case: 
![image](https://user-images.githubusercontent.com/8766776/149427733-e35fdaa6-d674-4df9-b097-99d8188506b3.png)

Reply:
I will use PackageIdentity as cache key.